### PR TITLE
add experimental Cody search page

### PR DESCRIPTION
**Pull request by sqs**
The Cody search page at `/search/cody` (behind the `cody` feature flag) accepts inputs such as `react class components` or `authentication changes` and converts them into Sourcegraph search query syntax.

This is an experimental feature and requires the credentials in the private Google doc at https://docs.google.com/document/d/1u7HYPmJFtDANtBgczzmAR0BmhM86drwDXCqx-F2jTEE/edit#.

image
image

## Test plan

n/a; experimental feature behind a feature flag

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| -0.02% (-0.43 kb) | 0.03% (+4.96 kb) | 0.05% (+5.39 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/sqs/cody-search-page/compare-report.html;tab=live_object) for a full comparison between the commits c0c8510 and ee8f304 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: client/web/src/enterprise/cody/api.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    const { sendMessage, lastMessage, readyState } = useWebSocket(authenticatedEndpointURL, {
+        reconnectAttempts: 3,
+        reconnectInterval: 500,
+        shouldReconnect: () => true,
+    })
+
+    return { sendMessage, lastMessage, readyState }
+}
+
+export function getCodyCompletionOneShot(prompt: string): Promise<string> {
```
Comment (novoselrok) [inline client/web/src/enterprise/cody/api.ts:41]: The completions API is also available on Sourcegraph instances through the `/.api/completions/stream` endpoint. Completions can be enabled through the site config with the appropriate API key.

It's a streaming endpoint, but if you only care about the final result, you can wait for the entire request to finish and extract the last `data: {"completion": "..."}` line.

Comment (sqs) [reply client/web/src/enterprise/cody/api.ts]: Nice. I just updated this PR to use that API.

File: package.json [hunk 2]
```diff
@@ -332,16 +341,7 @@
...
-    "yauzl": "^2.10.0",
-    "concurrently": "^7.6.0",
-    "@types/stopword": "^2.0.0",
-    "@types/stream-json": "^1.7.3",
-    "@vscode/vsce": "^2.18.0",
-    "@vitejs/plugin-react": "^3.1.0",
-    "vite": "^4.1.4",
-    "nodemon": "^2.0.20",
-    "@types/body-parser": "^1.19.2",
-    "@types/google-spreadsheet": "^3.3.1"
```
Comment (philipp-spiess) [inline package.json:344]: Were the `package.json` changes accidental? These are still used inside the cody client I think

**Review by philipp-spiess — APPROVED** [review]
This looks awesome! @fkling what about integrating this as an alternative search mode into the new search input. So instead of the history mode we'll also have an ✨ cody mode.  👀

Stamping to unblock, this is behind a feature flag anyways from what I can see. The `package.json` changes may look unrelated though.

**Review by novoselrok — APPROVED** [review]

File: package.json (hunk 2 above)
Comment (sqs) [reply package.json:344]: oops, I fixed this but must have forgotten to re-push

**Pull request merged**